client/client.go

```diff
@@ -388,19 +390,87 @@ func createClientWithKeyspace(
 	return c, nil
 }
 
-// NewClientWithKeyspaceName creates a client with context and the specified keyspace name.
-func NewClientWithKeyspaceName(
+// APIVersion is the API version the server and the client is using.
+// See more details in https://github.com/tikv/rfcs/blob/master/text/0069-api-v2.md#kvproto
+type APIVersion int
+
+// The API versions the client supports.
+const (
+	V1 APIVersion = iota
+	V1TTL
+	V2
+)
+
+// APIContext is the context for API version.
+type APIContext interface {
+	GetAPIVersionInfo() (apiVersion APIVersion, keyspaceName string)
+}
+
+type apiContextV1 struct{}
+
+// NewAPIContextV1 creates a API context for V1.
+func NewAPIContextV1() APIContext {
+	return &apiContextV1{}
+}
+
+// GetAPIVersionInfo returns the API version and the keyspace name.
+func (apiCtx *apiContextV1) GetAPIVersionInfo() (version APIVersion, keyspaceName string) {
+	return V1, ""
+}
+
+type apiContextV1TTL struct{}
+
+// NewAPIContextV1TTL creates a API context for V1TTL.
+func NewAPIContextV1TTL() APIContext {
+	return &apiContextV1TTL{}
+}
+
+// GetAPIVersionInfo returns the API version and the keyspace name.
+func (apiCtx *apiContextV1TTL) GetAPIVersionInfo() (version APIVersion, keyspaceName string) {
+	return V1TTL, ""
+}
+
+type apiContextV2 struct {
+	keyspaceName string
+}
+
+// NewAPIContextV2 creates a API context with the specified keyspace name for V2.
+func NewAPIContextV2(keyspaceName string) APIContext {
+	return &apiContextV2{keyspaceName: keyspaceName}
+}
+
+// GetAPIVersionInfo returns the API version and the keyspace name.
+func (apiCtx *apiContextV2) GetAPIVersionInfo() (version APIVersion, keyspaceName string) {
+	return V2, apiCtx.keyspaceName
+}
+
+// NewClientWithAPIContext creates a client according to the API context.
+func NewClientWithAPIContext(
+	ctx context.Context, apiCtx APIContext, svrAddrs []string,
+	security SecurityOption, opts ...ClientOption,
+) (Client, error) {
+	apiVersion, keyspaceName := apiCtx.GetAPIVersionInfo()
+	switch apiVersion {
+	case V1, V1TTL:
+		return NewClientWithContext(ctx, svrAddrs, security, opts...)
+	case V2:
+		if keyspaceName == "" {
+			keyspaceName = defaultKeyspaceName
+		}
+		return newClientWithKeyspaceName(ctx, keyspaceName, svrAddrs, security, opts...)
+	default:
+		return nil, errors.Errorf("[pd] invalid API version %d", apiVersion)
+	}
+}
+
+// newClientWithKeyspaceName creates a client with context and the specified keyspace name.
+func newClientWithKeyspaceName(
 	ctx context.Context, keyspaceName string, svrAddrs []string,
 	security SecurityOption, opts ...ClientOption,
 ) (Client, error) {
 	log.Info("[pd] create pd client with endpoints and keyspace",
 		zap.Strings("pd-address", svrAddrs), zap.String("keyspace-name", keyspaceName))
 
```

Comment:
How about move `if keyspaceName == ""` logic to here?
newClientWithKeyspaceName is only used for apiv2;
If someone else calls this function, there is no need to repeat the judgment logic

Reply:
https://github.com/tikv/pd/pull/6482#discussion_r1197502311 may be better?